### ISSUE TITLE
Decode vs. Display order encoding

### ISSUE THREAD

**Issue by lferraz** [NVIDIA/VideoProcessingFramework#212]
I know that packets in the encoded video can be "unsorted", so my question is when I use VPF  to encode videos, e.g. using `EncodeSingleSurface()`, can I get unsorted packets?

In my tests looks like `pts` and `dts` are always aligned (I hardcode them and the video looks ok), however is this always like this?

**Comment by rarzumanyan** [comment]
Hi @lferraz 

>can I get unsorted packets?

Not sure if I fully understand the question. Please correct me if I'm missing something.

Nvenc produces compliant Annex.B elementary bitstream. For encoded video sequences with B frames, decoding order is different from display order, but that's part of H.264 and H.265 standards and it's maintained by Nvenc.

You can find more information on this topic in `NVENC_VideoEncoder_API_ProgGuide.pdf` which is shipped as part of Video Codec SDK.

**Comment by lferraz** [comment]
HI @rarzumanyan , I will read that pdf. My point is that if the decoding order is different to the display one that order should be provided. 

Thanks for the answer

**Comment by lferraz** [comment]
from that pdf I found this:

If the client wants to send the input buffers in display order, it must set `enablePTD = 1`.
If the client wants to send the input buffers in encode order, it must set `enablePTD = 0`.

In VPF this is hardcoded `params.enablePTD = 1;` in `/PyNvCodec/TC/src/NvCodecCliOptions.cpp`
I also read that the sync encoding is not compatible with this parameter. I suppose internally it is disabled in this case (I couldn't find code in VPF to deal with this)

Then, looks like VPF can return encoded packets that are not in the display order when async encoding is enabled. So, how can I know the order if I do not get the pts and dts timestamps from the encoding?

**Comment by rarzumanyan** [comment]
@lferraz 

Sync / Async encode mode is selected here:
https://github.com/NVIDIA/VideoProcessingFramework/blob/906b6dc43e6be99284c24e382cf5fc93196d99c7/PyNvCodec/TC/src/NvCodecCliOptions.cpp#L515-L521

>I also read that the sync encoding is not compatible with this parameter. I suppose internally it is disabled in this case (I couldn't find code in VPF to deal with this)

Not sure if I fully understand this. I don't see anything in API Programming Guide telling that `enablePTD` is incompatible with sync Nvenc operational mode. Could you please reference particular section of the `NVENC_VideoEncoder_API_ProgGuide.pdf`?

**Comment by lferraz** [comment]
@rarzumanyan 
sorry, I found that in an old version of the API looking on the internet about info around the `enablePTD` parameter.
In the last version that paragraph was removed. Sorry, my mistake.

Anyway, I'd like to understand how to know the correct display order for the encoded frames.

**Comment by rarzumanyan** [comment]
@lferraz 

Since the `enablePTD = 1` is hard-coded, you just feed the encoder with video frames in display order.
Encoded packets you obtain from `EncodeSingleSurface()` are in order which allows compliant decoder to reconstruct them and present in display order the same way you've fed them to Nvenc.

Am I missing something?

**Comment by lferraz** [comment]
@rarzumanyan 
then I do not understand the diff between both modes. From the doc:
```
If the client wants to send the input buffers in display order, it must set enablePTD = 1. If
enablePTD is set to 1 the decision of determining the picture type will be taken by NVENCODE
API.
If the client wants to send the input buffers in encode order, it must set enablePTD = 0, and
must specify
NV_ENC_PIC_PARAMS::pictureType
NV_ENC_PIC_PARAMS_H264/NV_ENC_PIC_PARAMS_HEVC::displayPOCSyntax
NV_ENC_PIC_PARAMS_H264/NV_ENC_PIC_PARAMS_HEVC::refPicFlag
```
I assumed that with enablePTD==1 the encoder decides the best order to encode. While with enablePTD==0 the encoding order is the same as the input so, if the input buffers are in display order for sure will be encoded in the same order.

Probably I misunderstood something... I assumed that the `the picture type` mentioned in the doc also includes the ability to change the order of the frames.

**Comment by rarzumanyan** [comment]
@lferraz 

If `enablePTD == 1` it simplifies life but slightly increases memory consumption because Nvenc may require additional memory for frame buffering and re-ordering. It may also introduce some additional latency.

For latency-critical scenarios there's usually a bunch of settings allowing to squeeze a last bit of performance out of Nvenc such as:

1. Infinite GOP + Intra Refresh.
2. Turn off `enablePTD` because there's no frame reordering and all the frames except 1st are P frames.
3. Turn off multiple reference frames for better error resiliency.
4. Frequent SPS / PPS repetition.

**Comment by lferraz** [comment]
@rarzumanyan , 

thank you so much for the info.

My only concern is about reordering. How can I know the correct order when `enablePTD == 1` if I do not have the pts and dts?

**Comment by rarzumanyan** [comment]
@lferraz 

There's no such language as "Presentation Timestamp" and "Decode Timestamp" in H.264 and H.265 standards. 

**To my best knowledge** frame rate, duration and timestamp information is optional and is not required to reconstruct luma and chroma samples. Primary mean to determine picture output order is Picture Order Count value which is integral part of H.264 and H.265 standards.

P. S.
You can pass frame timestamp via `NV_ENC_PIC_PARAMS::inputTimeStamp` but as it's described in NvEncodeAPI.h this information will not be encoded in output bitstream.

>How can I know the correct order when enablePTD == 1 if I do not have the pts and dts?

Basically it's up to encoder to maintain correct frame order.
If you come across an issue with video frames being shuffled, that would require careful investigation.

**Comment by lferraz** [comment]
@rarzumanyan ,
thank you so much for your detailed and clear explanation. At this moment I couldn't see the shuffling issue but since I knew that the decoding and presentation order can be different I wanted to ask.

I will run some experiments to validate that they are perfectly ordered.

Thanks for your help!!!!!

**Comment by lferraz** [comment]
yes, but I think @rarzumanyan left it open because of informative reasons.
